modules/ximgproc/src/globalmatting.cpp

```diff
@@ -0,0 +1,568 @@
+	{
+	    if (trimap(y, x) != 128)
+	        continue;
+
+	    const cv::Vec3b &I = image(y, x);
+
+	    for (int j = y-r; j <= y+r; ++j)
+	        for (int i = x-r; i <= x+r; ++i)
+	        {
+	            if (i < 0 || i >= w || j < 0 || j >= h)
+	                continue;
+
+	            if (trimap(j, i) != 0 && trimap(j, i) != 255)
+	                continue;
+
+	            const cv::Vec3b &I2 = image(j, i);
+
+	            float pd = sqrt((float)(sqr(x - i) + sqr(y - j)));
+	            float cd = colorDist(I, I2);
+
+	            if (pd <= r && cd <= c)
+	            {
+	                if (trimap(j, i) == 0)
+	                    trimap(y, x) = 1;
+	                else if (trimap(j, i) == 255)
+	                    trimap(y, x) = 254;
+	            }
+	        }
+	}
+
+    for (int x = 0; x < trimap.cols; ++x)
+	for (int y = 0; y < trimap.rows; ++y)
+	{
+	    if (trimap(y, x) == 1)
+	        trimap(y, x) = 0;
+	    else if (trimap(y, x) == 254)
+	        trimap(y, x) = 255;
+
+	}
+}
+
+// erode foreground and background regions to increase the size of unknown region
+void GlobalMatting::erodeFB(cv::Mat_<uchar> &trimap, int r)
+{
+    int w = trimap.cols;
+    int h = trimap.rows;
+
+    cv::Mat_<uchar> foreground(trimap.size(), (uchar)0);
+    cv::Mat_<uchar> background(trimap.size(), (uchar)0);
+
+    for (int y = 0; y < h; ++y)
+	for (int x = 0; x < w; ++x)
+	{
+	    if (trimap(y, x) == 0)
+	        background(y, x) = 1;
+	    else if (trimap(y, x) == 255)
+	        foreground(y, x) = 1;
+	}
+
+
+    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(r, r));
+
+    cv::erode(background, background, kernel);
+    cv::erode(foreground, foreground, kernel);
+
+    for (int y = 0; y < h; ++y)
+	for (int x = 0; x < w; ++x)
+	{
+	    if (background(y, x) == 0 && foreground(y, x) == 0)
+	        trimap(y, x) = 128;
+	}
+}
+
+
+void GlobalMatting::calculateAlphaPatchMatch(const cv::Mat_<cv::Vec3b> &image,
+	const cv::Mat_<uchar> &trimap,
+	const std::vector<cv::Point> &foregroundBoundary,
+	const std::vector<cv::Point> &backgroundBoundary,
+	std::vector<std::vector<Sample> > &samples)
+{
+    int w = image.cols;
+    int h = image.rows;
+
+    samples.resize(h, std::vector<Sample>(w));
+
+    for (int y = 0; y < h; ++y)
+	for (int x = 0; x < w; ++x)
+	{
+	    if (trimap(y, x) == 128)
+	    {
+	        cv::Point p(x, y);
+
+	        samples[y][x].fi = rand() % foregroundBoundary.size();
+	        samples[y][x].bj = rand() % backgroundBoundary.size();
+	        samples[y][x].df = nearestDistance(foregroundBoundary, p);
+	        samples[y][x].db = nearestDistance(backgroundBoundary, p);
+	        samples[y][x].cost = FLT_MAX;
+	    }
+	}
+
+    std::vector<cv::Point> coords(w * h);
+    for (int y = 0; y < h; ++y)
+	for (int x = 0; x < w; ++x)
+	    coords[x + y * w] = cv::Point(x, y);
+
+    for (int iter = 0; iter < 10; ++iter)
+    {
+	// propagation
+	std::random_shuffle(coords.begin(), coords.end());
```

Comment:
Avoid using of std::random functions.
- their implementations may be not thread safe
- so their results are not reproducible (hard to reproduce test failures)

Use OpenCV's functions instead (they are thread-safe with thread-local RNG state, which is propagated through `cv::parallel_for()`).